File: apps/android/app/src/main/kotlin/com/wordocious/app/ui/AnnouncementsBanner.kt
package com.wordocious.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wordocious.app.data.SettingsPref
import com.wordocious.app.data.SupabaseConfig
import com.wordocious.app.ui.theme.Nunito
import com.wordocious.app.ui.theme.WTheme
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.Serializable

// Native reader for admin-authored announcements (web/iOS parity). RLS
// serves only active + unexpired rows; dismissals persist per-id.
@Composable
fun AnnouncementsBanner() {
    var current by remember { mutableStateOf<AnnouncementRow?>(null) }

    LaunchedEffect(Unit) {
        val rows = runCatching {
            SupabaseConfig.client.postgrest["announcements"]
                .select(Columns.raw("id, title, body")) {
                    order("created_at", Order.DESCENDING)
                    limit(5)
                }
                .decodeList<AnnouncementRow>()
        }.getOrDefault(emptyList())
        val dismissed = SettingsPref.get("dismissed-announcements", "").split(",").toSet()
        current = rows.firstOrNull { it.id !in dismissed }
    }

    val a = current ?: return
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(WTheme.surface, RoundedCornerShape(16.dp))
            .border(1.5.dp, Color(0xFFC4B5FD), RoundedCornerShape(16.dp))
            .padding(12.dp),
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Text("📣", fontSize = 13.sp)
        Column(Modifier.weight(1f)) {
            Text(
                a.title, fontSize = 13.sp, fontWeight = FontWeight.Black,
                style = TextStyle(
                    brush = Brush.linearGradient(listOf(Color(0xFF7C3AED), Color(0xFFEC4899))),
                    fontFamily = Nunito,
                ),
            )
            Text(a.body, fontSize = 11.sp, fontWeight = FontWeight.Bold, color = WTheme.textMuted, fontFamily = Nunito)
        }
        Text("✕", fontSize = 12.sp, color = WTheme.textMuted, modifier = Modifier.clickableNoRipple {
            val dismissed = SettingsPref.get("dismissed-announcements", "")
            SettingsPref.set("dismissed-announcements", if (dismissed.isEmpty()) a.id else "$dismissed,${a.id}")
            current = null
        })
    }
}

@Serializable
data class AnnouncementRow(val id: String, val title: String, val body: String)

File: apps/ios/Wordocious/Sources/AnnouncementsBanner.swift
import SwiftUI

/// Native reader for the admin-authored announcements table (web parity:
/// components/ui/announcements-banner.tsx). RLS serves only active +
/// unexpired rows; dismissals are per-announcement in UserDefaults.
struct AnnouncementsBanner: View {
    struct Announcement: Decodable, Identifiable {
        let id: String
        let title: String
        let body: String
    }

    @State private var current: Announcement?
    private static let dismissedKey = "dismissed-announcements"

    var body: some View {
        Group {
            if let a = current {
                HStack(alignment: .top, spacing: 10) {
                    Image(systemName: "megaphone.fill")
                        .font(.system(size: 13)).foregroundStyle(Color(hex: 0x7C3AED))
                        .padding(.top, 2)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(a.title)
                            .font(Brand.font(13, .black))
                            .foregroundStyle(LinearGradient(colors: [Color(hex: 0x7C3AED), Color(hex: 0xEC4899)], startPoint: .leading, endPoint: .trailing))
                        Text(a.body).font(Brand.font(11, .bold)).foregroundStyle(Theme.textMuted)
                    }
                    Spacer(minLength: 0)
                    Button { dismiss(a) } label: {
                        Image(systemName: "xmark").font(.system(size: 11, weight: .bold))
                            .foregroundStyle(Theme.textMuted)
                    }.buttonStyle(.plain)
                }
                .padding(12)
                .background(RoundedRectangle(cornerRadius: 16).fill(Theme.surface))
                .overlay(RoundedRectangle(cornerRadius: 16).stroke(Color(hex: 0xC4B5FD), lineWidth: 1.5))
            }
        }
        .task { await load() }
    }

    private func dismissedIds() -> Set<String> {
        Set(UserDefaults.standard.stringArray(forKey: Self.dismissedKey) ?? [])
    }

    private func dismiss(_ a: Announcement) {
        var ids = dismissedIds(); ids.insert(a.id)
        UserDefaults.standard.set(Array(ids), forKey: Self.dismissedKey)
        withAnimation { current = nil }
    }

    private func load() async {
        let rows: [Announcement]? = try? await AuthService.shared.client
            .from("announcements")
            .select("id, title, body")
            .order("created_at", ascending: false)
            .limit(5)
            .execute().value
        let dismissed = dismissedIds()
        current = rows?.first { !dismissed.contains($0.id) }
    }
}
